File: frontend/src/api/canvases.ts
import client from "./client";
import type {
  CanvasDetail,
  CanvasEdge,
  CanvasNode,
  CanvasTaskCreateResponse,
  CanvasTaskPayload,
  UserCanvasListResponse,
  UserCanvasSummary,
} from "@/types";

export function listCanvases(): Promise<UserCanvasListResponse> {
  return client.get("/canvases");
}

export function getDefaultCanvasName(date = new Date()): string {
  const year = String(date.getFullYear()).slice(-2);
  const month = String(date.getMonth() + 1).padStart(2, "0");
  const day = String(date.getDate()).padStart(2, "0");
  return `新画板-${year}${month}${day}`;
}

export function createCanvas(name: string = getDefaultCanvasName()): Promise<UserCanvasSummary> {
  return client.post("/canvases", { name });
}

export function getCanvas(projectId: string): Promise<CanvasDetail> {
  return client.get(`/canvases/${projectId}`);
}

export function updateCanvas(projectId: string, data: { name?: string; viewport_x?: number; viewport_y?: number; zoom?: number }): Promise<UserCanvasSummary> {
  return client.patch(`/canvases/${projectId}`, data);
}

export function deleteCanvas(projectId: string): Promise<void> {
  return client.delete(`/canvases/${projectId}`);
}

export function updateCanvasViewport(projectId: string, data: { viewport_x: number; viewport_y: number; zoom: number }): Promise<UserCanvasSummary> {
  return client.patch(`/canvases/${projectId}/viewport`, data);
}

export function updateCanvasNode(projectId: string, nodeId: number, data: Partial<Pick<CanvasNode, "x" | "y" | "width" | "height" | "z_index" | "content">>): Promise<CanvasNode> {
  return client.patch(`/canvases/${projectId}/nodes/${nodeId}`, data);
}

export function updateCanvasNodesBatch(projectId: string, nodes: Array<Partial<Pick<CanvasNode, "x" | "y" | "width" | "height" | "z_index" | "content">> & { id: number }>): Promise<{ nodes: CanvasNode[] }> {
  return client.patch(`/canvases/${projectId}/nodes/batch`, { nodes });
}

export function deleteCanvasNode(projectId: string, nodeId: number): Promise<void> {
  return client.delete(`/canvases/${projectId}/nodes/${nodeId}`);
}

export function updateCanvasEdge(projectId: string, edgeId: number, data: { is_collapsed?: boolean }): Promise<CanvasEdge> {
  return client.patch(`/canvases/${projectId}/edges/${edgeId}`, data);
}

export function createCanvasNode(projectId: string, data: {
  node_type: "text" | "image";
  content?: string;
  image_url?: string;
  x: number;
  y: number;
  width?: number;
  height?: number;
}): Promise<CanvasNode> {
  return client.post(`/canvases/${projectId}/nodes`, data);
}

export function createCanvasTask(projectId: string, data: CanvasTaskPayload): Promise<CanvasTaskCreateResponse> {
  return client.post(`/canvases/${projectId}/tasks`, {
    ...data,
    source: data.source || "web",
  });
}

File: backend/app/api/canvases.py
import logging

from fastapi import APIRouter, Depends, HTTPException, Response, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_user
from app.database import get_db
from app.models.user import User
from app.schemas.canvas import (
    CanvasCreate,
    CanvasDetail,
    CanvasEdgeOut,
    CanvasEdgeUpdate,
    CanvasFreeNodeCreate,
    CanvasListResponse,
    CanvasNodeBatchUpdate,
    CanvasNodeBatchUpdateResponse,
    CanvasNodeOut,
    CanvasNodeUpdate,
    CanvasSummary,
    CanvasTaskCreate,
    CanvasTaskCreateResponse,
    CanvasUpdate,
    CanvasViewportUpdate,
)
from app.services.business_id_service import task_external_id, user_external_id
from app.services.canvas_service import (
    create_canvas_generation_tasks,
    create_canvas_free_node,
    create_user_canvas,
    delete_canvas_node,
    delete_user_canvas,
    get_canvas_detail,
    list_user_canvases,
    update_canvas_node,
    update_canvas_nodes_batch,
    update_canvas_edge,
    update_canvas_viewport,
    update_user_canvas,
)
from app.services.external_api_config_service import (
    SCENE_INPAINT,
    get_default_generation_model_key,
    require_scene_config,
)
from app.services.task_service import (
    mark_tasks_dispatched,
    mark_tasks_enqueue_failed,
    mark_tasks_queued,
)

router = APIRouter(prefix="/api/canvases", tags=["无限画布"])
canvas_logger = logging.getLogger("app.canvas")


@router.get("", response_model=CanvasListResponse)
def list_canvases(
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return list_user_canvases(db, user.id)


@router.post("", response_model=CanvasSummary, status_code=status.HTTP_201_CREATED)
def create_canvas(
    body: CanvasCreate,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return create_user_canvas(db, user.id, body.name)


@router.get("/{project_id}", response_model=CanvasDetail)
def get_canvas(
    project_id: str,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return get_canvas_detail(db, user.id, project_id, allow_admin_read=user.role in {"admin", "superadmin"})


@router.patch("/{project_id}", response_model=CanvasSummary)
def update_canvas(
    project_id: str,
    body: CanvasUpdate,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return update_user_canvas(
        db,
        user.id,
        project_id,
        name=body.name,
        viewport_x=body.viewport_x,
        viewport_y=body.viewport_y,
        zoom=body.zoom,
    )


@router.delete("/{project_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_canvas(
    project_id: str,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    delete_user_canvas(db, user.id, project_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.patch("/{project_id}/viewport", response_model=CanvasSummary)
def update_viewport(
    project_id: str,
    body: CanvasViewportUpdate,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return update_canvas_viewport(
        db,
        user.id,
        project_id,
        viewport_x=body.viewport_x,
        viewport_y=body.viewport_y,
        zoom=body.zoom,
    )


@router.patch("/{project_id}/nodes/batch", response_model=CanvasNodeBatchUpdateResponse)
def update_nodes_batch(
    project_id: str,
    body: CanvasNodeBatchUpdate,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return update_canvas_nodes_batch(db, user.id, project_id, updates=body.nodes)


@router.patch("/{project_id}/nodes/{node_id}", response_model=CanvasNodeOut)
def update_node(
    project_id: str,
    node_id: int,
    body: CanvasNodeUpdate,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return update_canvas_node(
        db,
        user.id,
        project_id,
        node_id,
        x=body.x,
        y=body.y,
        width=body.width,
        height=body.height,
        z_index=body.z_index,
        content=body.content,
    )


@router.delete("/{project_id}/nodes/{node_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_node(
    project_id: str,
    node_id: int,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    delete_canvas_node(db, user.id, project_id, node_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.patch("/{project_id}/edges/{edge_id}", response_model=CanvasEdgeOut)
def update_edge(
    project_id: str,
    edge_id: int,
    body: CanvasEdgeUpdate,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return update_canvas_edge(
        db,
        user.id,
        project_id,
        edge_id,
        is_collapsed=body.is_collapsed,
    )


@router.post("/{project_id}/nodes", response_model=CanvasNodeOut, status_code=status.HTTP_201_CREATED)
def create_free_node(
    project_id: str,
    body: CanvasFreeNodeCreate,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return create_canvas_free_node(
        db,
        user.id,
        project_id,
        node_type=body.node_type,
        content=body.content,
        image_url=body.image_url,
        x=body.x,
        y=body.y,
        width=body.width,
        height=body.height,
    )


@router.post("/{project_id}/tasks", response_model=CanvasTaskCreateResponse, status_code=status.HTTP_201_CREATED)
def create_canvas_task(
    project_id: str,
    body: CanvasTaskCreate,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    if body.mode == "inpaint":
        require_scene_config(db, SCENE_INPAINT)
        task_model = SCENE_INPAINT
        resolved_resolution = body.resolution
    else:
        task_model = body.model.strip() or get_default_generation_model_key(db)
        require_scene_config(db, task_model)
        resolved_resolution = "" if task_model == "banana" else body.resolution

    try:
        from app.workers.generation import dispatch_generation_task, get_generation_dispatch_mode

        dispatch_mode = get_generation_dispatch_mode()
    except RuntimeError as exc:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail=str(exc),
        ) from exc

    tasks, nodes = create_canvas_generation_tasks(
        db,
        user.id,
        project_id,
        model=task_model,
        source=body.source,
        mode=body.mode,
        prompt=body.prompt,
        num_images=body.num_images,
        size=body.size,
        resolution=resolved_resolution,
        custom_size=body.custom_size,
        reference_images=body.reference_images,
        source_node_ids=body.source_node_ids,
        source_image=body.source_image,
        mask_image=body.mask_image,
        x=body.x,
        y=body.y,
        width=body.width,
        height=body.height,
    )

    dispatched_task_ids: list[int] = []
    try:
        for task in tasks:
            actual_dispatch_mode = dispatch_generation_task(task.id)
            dispatched_task_ids.append(task.id)
            canvas_logger.info(
                "canvas task dispatched",
                extra={
                    "event": "canvas.task.dispatch.sent",
                    "user_id": user_external_id(user),
                    "project_id": project_id,
                    "task_id": task_external_id(task),
                    "dispatch_mode": actual_dispatch_mode,
                },
            )
        mark_tasks_dispatched(db, dispatched_task_ids)
        if dispatch_mode == "celery":
            mark_tasks_queued(db, dispatched_task_ids)
    except Exception as exc:
        failed_task_ids = [task.id for task in tasks if task.id not in set(dispatched_task_ids)]
        mark_tasks_enqueue_failed(db, failed_task_ids, error_message=str(exc))
        if not dispatched_task_ids:
            raise HTTPException(
                status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
                detail="任务队列暂不可用，请稍后重试",
            ) from exc

    task_ids = [task_external_id(task) for task in tasks]
    return CanvasTaskCreateResponse(task_id=task_ids[0] if task_ids else None, task_ids=task_ids, nodes=nodes)
